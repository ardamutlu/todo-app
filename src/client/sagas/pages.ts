import { END } from 'redux-saga';
import { put, call, takeLatest } from 'redux-saga/effects';
import {
  setEnv,
  resetPageStatus,
  loadAppProcess as LoadAppProcess,
  loadAppProcessSuccess,
  loadHomePage as LoadHomePage,
  loadHomePageSuccess,
  loadDocumentationPage as LoadDocumentationPage,
  loadDocumentationPageSuccess,
  LOAD_APP_PROCESS,
  LOAD_HOME_PAGE,
  LOAD_DOCUMENTATION_PAGE
} from '../actions/pages';
import { getAllTodoRequest } from '../actions/todo';

// e.g. write common processing to be performed on all pages
// don't call `stopSaga`
function* appProcess(actions: ReturnType<typeof LoadAppProcess>) {
  yield put(setEnv(process.env.NODE_ENV || 'development'));
  yield put(loadAppProcessSuccess());
}

// if you run async process, you have to change the code like below
// function* appProcess(actions: ReturnType<typeof LoadAppProcess>) {
//   try {
//     // async

//     yield put(loadAppProcessSuccess());
//   } catch (err) {
//     yield put(loadAppProcessFailure(err));
//   } finally {
//     if (!process.env.IS_BROWSER) {
//       yield call(stopSaga);
//     }
//   }
// }

function* loadHomePage(actions: ReturnType<typeof LoadHomePage>) {
  yield changePage();
  // When homepage load get all todo's
  yield put(getAllTodoRequest());
  yield put(loadHomePageSuccess());

  if (!process.env.IS_BROWSER) {
    yield call(stopSaga);
  }
}

function* loadDocumentationPage(actions: ReturnType<typeof LoadDocumentationPage>) {
  yield changePage();
  yield put(loadDocumentationPageSuccess());
  if (!process.env.IS_BROWSER) {
    yield call(stopSaga);
  }
}

function* stopSaga() {
  yield put(END);
}

function* changePage() {
  // don't need to call resetPageStatus because baseUrl is required by fetch on Node.js environment,
  // also state has already been initialized at this time
  if (process.env.IS_BROWSER) {
    yield put(resetPageStatus());
  }
}

export function* pagesProcess() {
  yield takeLatest(LOAD_APP_PROCESS, appProcess);
  yield takeLatest(LOAD_HOME_PAGE, loadHomePage);
  yield takeLatest(LOAD_DOCUMENTATION_PAGE, loadDocumentationPage);
}
