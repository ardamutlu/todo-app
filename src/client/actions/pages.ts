import { State } from '../reducers/pages';

export type Actions =
  | ReturnType<typeof setEnv>
  | ReturnType<typeof setIsMobile>
  | ReturnType<typeof setBaseUrl>
  | ReturnType<typeof resetPageStatus>
  | ReturnType<typeof loadAppProcess>
  | ReturnType<typeof loadAppProcessSuccess>
  | ReturnType<typeof loadAppProcessFailure>
  | ReturnType<typeof loadHomePage>
  | ReturnType<typeof loadHomePageSuccess>
  | ReturnType<typeof loadHomePageFailure>
  | ReturnType<typeof loadDocumentationPage>
  | ReturnType<typeof loadDocumentationPageSuccess>
  | ReturnType<typeof loadDocumentationPageFailure>;

export const SET_ENV = 'SET_ENV';
export const setEnv = (env: State['env']) =>
  ({
    type: SET_ENV,
    payload: {
      env
    }
  } as const);

export const SET_IS_MOBILE = 'SET_IS_MOBILE';
export const setIsMobile = (isMobile: State['isMobile']) =>
  ({
    type: SET_IS_MOBILE,
    payload: {
      isMobile
    }
  } as const);

export const SET_BASE_URL = 'SET_BASE_URL';
export const setBaseUrl = (baseUrl: State['baseUrl']) =>
  ({
    type: SET_BASE_URL,
    payload: {
      baseUrl
    }
  } as const);

export const RESET_PAGES_STATUS = 'RESET_PAGES_STATUS';
export const resetPageStatus = () =>
  ({
    type: RESET_PAGES_STATUS
  } as const);

export const LOAD_APP_PROCESS = 'LOAD_APP_PROCESS';
export const loadAppProcess = () =>
  ({
    type: LOAD_APP_PROCESS
  } as const);

export const LOAD_APP_PROCESS_SUCCESS = 'LOAD_APP_PROCESS_SUCCESS';
export const loadAppProcessSuccess = () =>
  ({
    type: LOAD_APP_PROCESS_SUCCESS
  } as const);

export const LOAD_APP_PROCESS_FAILURE = 'LOAD_APP_PROCESS_FAILURE';
export const loadAppProcessFailure = (err: Error) =>
  ({
    type: LOAD_APP_PROCESS_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_HOME_PAGE = 'LOAD_HOME_PAGE';
export const loadHomePage = () =>
  ({
    type: LOAD_HOME_PAGE
  } as const);

export const LOAD_HOME_PAGE_SUCCESS = 'LOAD_HOME_PAGE_SUCCESS';
export const loadHomePageSuccess = () =>
  ({
    type: LOAD_HOME_PAGE_SUCCESS
  } as const);

export const LOAD_HOME_PAGE_FAILURE = 'LOAD_HOME_PAGE_FAILURE';
export const loadHomePageFailure = (err: Error) =>
  ({
    type: LOAD_HOME_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_DOCUMENTATION_PAGE = 'LOAD_DOCUMENTATION_PAGE';
export const loadDocumentationPage = () =>
  ({
    type: LOAD_DOCUMENTATION_PAGE
  } as const);

export const LOAD_DOCUMENTATION_PAGE_SUCCESS = 'LOAD_DOCUMENTATION_PAGE_SUCCESS';
export const loadDocumentationPageSuccess = () =>
  ({
    type: LOAD_DOCUMENTATION_PAGE_SUCCESS
  } as const);

export const LOAD_DOCUMENTATION_PAGE_FAILURE = 'LOAD_DOCUMENTATION_PAGE_FAILURE';
export const loadDocumentationPageFailure = (err: Error) =>
  ({
    type: LOAD_DOCUMENTATION_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);
