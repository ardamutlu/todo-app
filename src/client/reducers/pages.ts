import * as actions from '../actions/pages';

export type State = {
  isLoadingCompletion: boolean;
  error: Error | null;
  baseUrl: string;
  env: typeof process.env.NODE_ENV;
  isMobile: boolean;
};

export const initialState: State = {
  isLoadingCompletion: false,
  error: null,
  baseUrl: '',
  env: 'development',
  isMobile: false
};

export const reducer = (state: State = initialState, action: actions.Actions): State => {
  switch (action.type) {
    case actions.SET_ENV:
      return { ...state, env: action.payload.env };
    case actions.SET_IS_MOBILE:
      return { ...state, isMobile: action.payload.isMobile };
    case actions.RESET_PAGES_STATUS:
      return initialState;
    case actions.SET_BASE_URL:
      return { ...state, baseUrl: action.payload.baseUrl };
    case actions.LOAD_HOME_PAGE:
    case actions.LOAD_DOCUMENTATION_PAGE:
      return { ...state, isLoadingCompletion: false };
    case actions.LOAD_HOME_PAGE_SUCCESS:
    case actions.LOAD_DOCUMENTATION_PAGE_SUCCESS:
      return { ...state, isLoadingCompletion: true };
    case actions.LOAD_HOME_PAGE_FAILURE:
    case actions.LOAD_DOCUMENTATION_PAGE_FAILURE:
      return { ...state, isLoadingCompletion: true, error: action.payload.err };
    default:
      return state;
  }
};
