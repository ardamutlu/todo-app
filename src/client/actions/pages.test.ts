import * as actions from './pages';

test('should create an action to set a base-url', () => {
  expect(actions.setBaseUrl('url')).toMatchSnapshot();
});

test('should create an action to reset state', () => {
  expect(actions.resetPageStatus()).toMatchSnapshot();
});

test('should create an action to load app process', () => {
  expect(actions.loadAppProcess()).toMatchSnapshot();
});

test('should create a success action to load app process', () => {
  expect(actions.loadAppProcessSuccess()).toMatchSnapshot();
});

test('should create a failure action to load app process', () => {
  expect(actions.loadAppProcessFailure(new Error())).toMatchSnapshot();
});

test('should create an action to load Home page', () => {
  expect(actions.loadHomePage()).toMatchSnapshot();
});

test('should create a success action to load Home page', () => {
  expect(actions.loadHomePageSuccess()).toMatchSnapshot();
});

test('should create a failure action to load Home page', () => {
  expect(actions.loadHomePageFailure(new Error())).toMatchSnapshot();
});

test('should create an action to load Documentation page', () => {
  expect(actions.loadDocumentationPage()).toMatchSnapshot();
});

test('should create a success action to load Documentation page', () => {
  expect(actions.loadDocumentationPageSuccess()).toMatchSnapshot();
});

test('should create a failure action to load Documentation page', () => {
  expect(actions.loadDocumentationPageFailure(new Error())).toMatchSnapshot();
});
