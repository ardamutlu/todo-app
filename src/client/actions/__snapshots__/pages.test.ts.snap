// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a failure action to load Home page 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "LOAD_HOME_PAGE_FAILURE",
}
`;

exports[`should create a failure action to load app process 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "LOAD_APP_PROCESS_FAILURE",
}
`;

exports[`should create a success action to load Home page 1`] = `
Object {
  "type": "LOAD_HOME_PAGE_SUCCESS",
}
`;

exports[`should create a success action to load app process 1`] = `
Object {
  "type": "LOAD_APP_PROCESS_SUCCESS",
}
`;

exports[`should create an action to load Home page 1`] = `
Object {
  "type": "LOAD_HOME_PAGE",
}
`;

exports[`should create an action to load app process 1`] = `
Object {
  "type": "LOAD_APP_PROCESS",
}
`;

exports[`should create an action to reset state 1`] = `
Object {
  "type": "RESET_PAGES_STATUS",
}
`;

exports[`should create an action to set a base-url 1`] = `
Object {
  "payload": Object {
    "baseUrl": "url",
  },
  "type": "SET_BASE_URL",
}
`;
